ofrak_core/pytest_ofrak/utils.py

```diff
@@ -0,0 +1,62 @@
+import functools
+import inspect
+
+
+def auto_validate_state(validator_function):
+    """
+    Class decorator for services or other objects with complex internal state. Used to create
+    classes with methods that run a state validation function after every (public) method call.
+
+    Public methods are detected as any callable whose name does not start with an underscore ("_").
+
+    Example usage:
+    def validator_function(class_instance):
+        # validation logic
+        ...
+
+
+    @auto_validate_state(validator_function)
+    class AutoValidatingImplementation(ClassWithInternalState):
+        pass
+
+    An instance of AutoValidatingImplementation now runs validator_function on itself after every
+    method call.
+
+    :param validator_function:
+
+    :return:
+    """
```

Comment:
One thing that could be clearer in this docstring is that the validator function has to raise an exception if there is a problem. I could see a situation in which a user might think the validator has to return true or false. This should be an easy, minor clarification to make.